File: 2021/day_2/main.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

type command struct {
	direction string
	amount    int
}

func readFile(fileName string) ([]string, error) {
	data, err := os.ReadFile(fileName)
	if err != nil {
		return nil, err
	}
	parsed := string(data)
	split := strings.Split(parsed, "\n")
	return split, nil
}

func parseLines(lines []string) []command {
	result := []command{}
	for _, line := range lines {
		splitted := strings.Split(line, " ")
		amount, err := strconv.Atoi(splitted[1])
		if err != nil {
			panic(err)
		}
		result = append(result, command{
			direction: splitted[0], amount: amount,
		})
	}
	return result
}

func executeCommandsOne(commands []command) (int, int) {
	depth := 0
	length := 0
	for _, c := range commands {
		if c.direction == "forward" {
			length += c.amount
		} else if c.direction == "down" {
			depth += c.amount
		} else if c.direction == "up" {
			depth -= c.amount
		} else {
			panic(fmt.Sprintf("Unexpected command %s", c.direction))
		}
	}
	return depth, length
}

func executeCommandsTwo(commands []command) (int, int) {
	depth, length, aim := 0, 0, 0
	for _, c := range commands {
		if c.direction == "forward" {
			length += c.amount
			depth += aim * c.amount
		} else if c.direction == "down" {
			aim += c.amount
		} else if c.direction == "up" {
			aim -= c.amount
		} else {
			panic(fmt.Sprintf("Unexpected command %s", c.direction))
		}
	}
	return depth, length
}

func main() {
	filename := "input.txt"
	lines, err := readFile(filename)
	if err != nil {
		lines, err = readFile(fmt.Sprintf("2021/day_2/%s", filename))
		if err != nil {
			panic(err)
		}
	}
	parsedLines := parseLines(lines)
	// depth, length := executeCommandsOne(parsedLines) // Part 1
	depth, length := executeCommandsTwo(parsedLines) // Part 2
	fmt.Printf("We went %dm deep and %dm far, totalling %d", depth, length, depth*length)
}

File: 2021/day_2/main.rs
use std::fs;

#[derive(Debug)]
struct Command {
    amount: i32,
    direction: String,
}

fn read_file(path: &str) -> Vec<String> {
    let parsed = fs::read_to_string(path).expect("Could not read file");
    let split_lines = parsed.lines();
    split_lines.map(|s| s.to_string()).collect()
}

fn parse_lines(lines: &Vec<String>) -> Vec<Command> {
    let mut result = Vec::<Command>::new();
    for l in lines {
        let splitted: Vec<String> = l.split_ascii_whitespace().map(|s| s.to_string()).collect();
        result.push(Command {
            amount: splitted[1].parse().unwrap(),
            direction: (*splitted[0]).to_string(),
        });
    }
    result
}

fn execute_commands_one(commands: &Vec<Command>) -> (i32, i32) {
    let (mut depth, mut distance) = (0, 0);
    for c in commands {
        match c.direction.as_str() {
            "forward" => distance += c.amount,
            "down" => depth += c.amount,
            "up" => depth -= c.amount,
            _ => panic!("Cannot parse command {:?}", c),
        };
    }
    (depth, distance)
}

fn execute_commands_two(commands: &Vec<Command>) -> (i32, i32) {
    let (mut aim, mut depth, mut distance) = (0, 0, 0);
    for c in commands {
        match c.direction.as_str() {
            "forward" => {
                distance += c.amount;
                depth += aim * c.amount;
            }
            "down" => aim += c.amount,
            "up" => aim -= c.amount,
            _ => panic!("Cannot parse command {:?}", c),
        };
    }
    (depth, distance)
}

fn main() {
    let filename = "2021/day_2/input.txt";
    let lines = read_file(filename);
    let parsed_commands = parse_lines(&lines);
    // let (depth, distance) = execute_commands_one(&parsed_commands); // Part 1
    let (depth, distance) = execute_commands_two(&parsed_commands); // Part 2
    println!(
        "We went {depth}m deep and {length}m far, totalling {total}",
        depth = depth,
        length = distance,
        total = depth * distance
    );
}
